snf-cyclades-app/synnefo/api/keypairs.py

```diff
@@ -0,0 +1,157 @@
+# Copyright (C) 2010-2016 GRNET S.A.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+
+from django.conf import settings
+from django.conf.urls import patterns
+from django.http import HttpResponse, HttpResponseServerError
+
+from snf_django.lib.astakos import get_user
+from snf_django.lib import api
+from snf_django.lib.api import utils, faults
+
+from synnefo.userdata.models import PublicKeyPair
+from synnefo.userdata.util import generate_keypair, SUPPORT_GENERATE_KEYS
+from synnefo.api import util
+from synnefo.db import transaction
+
+from logging import getLogger
+
+log = getLogger(__name__)
+
+urlpatterns = patterns(
+    'synnefo.api.keypairs',
+    (r'^(?:/|.json|.xml)?$', 'demux'),
+    (r'^/([\w-]+)(?:/|.json|.xml)?$', 'keypair_demux'),
+)
+
+
+def keypair_to_dict(keypair, details=False):
+    d = {
+        'name': keypair.name,
+        'fingerprint': keypair.fingerprint,
+        'public_key': keypair.content
+    }
+    if details is True:
+        d['deleted'] = keypair.deleted
+        d['id'] = keypair.id
+        d['created_at'] = keypair.created_at.isoformat()
+        d['deleted_at'] = None
+        if keypair.deleted_at is not None:
+            d['deleted_at'] = keypair.deleted_at.isoformat()
+        d['updated_at'] = None
+        if keypair.updated_at is not None:
+            d['updated_at'] = keypair.updated_at.isoformat()
+    return {'keypair': d}
```

Comment:
create_new_keypair() is not using keypair_to_dict(). The key creation method has an invalid response. Do a POST and check the response.